Update README for Code Pattern

ajbozarth opened the pull request:
## README Updates

- Added Code Pattern Title
- Added Code Pattern Overview
- Added Code Pattern Goals
- Added Flow and Architecture Diagram rough draft
- Added Steps ToC
- Added (Optional) API Step
- Updated icons README with License
- Made small text and formatting updates

## Code Pattern Documentation to be added to wiki

### Code Pattern Long Name (wiki: Short Description)
Use an open source object detector deep learning model to display and filter objects recognized in an image in a web application

### Code Pattern Description (wiki: Introduction)
The IBM Model Asset eXchange (MAX) has given application developers without data science experience easy access to prebuilt machine learning models. This code pattern shows how to create a simple web application to visualize the text output of a MAX model. The web app uses the Object Detector from MAX and creates a simple web UI that displays bounding boxes around detected objects in an image and lets you filter the objects based on their label and probable accuracy given by the model.

ajbozarth commented:
I'll being writing the remaining sections tomorrow but I wanted feedback on what I have.

Feel free to give feedback on any part of the README (even the merged parts) in this PR. This PR will contain the remaining README updates prior to submitting a code pattern proposal and cutting the v1.0 release of the web app.

@MLnick and @djalova I want your input the most on this.

@ptitzler and @gdequeiroz feel free to give input as well.

Also remember this will continue to be updated and edited during the code pattern process so we don't need to perfect everything here, I just want to make sure it doesn't have any errors and makes sense.

MLnick commented on README.md:
```diff
@@ -1,8 +1,24 @@
 
-# Object Detector Web App
+# Create a web app to visually interact with objects detected with machine learning
+
```
I guess this will be updated later perhaps, but I think the opening paragraphs should be more descriptive (ref: https://github.com/IBM/MAX-Image-Caption-Generator-Web-App/#create-a-web-app-to-interact-with-machine-learning-generated-image-captions)

ajbozarth replied:
yeah the one line will be replaced with multiple paragraphs today, that's the one WIP task, I just wanted eyes on the rest of the changes while I'm working on it today.

ajbozarth commented:
I've written up the two descriptive parts for the code pattern that don't go in the readme. If everyone could also take a look at these before I put them in the wiki. I'll also update the PR description with them. Also note that I based these on the ones from the previous web app code pattern (like copy-paste and update to match the differences in the models and web apps).

### Code Pattern Long Name (wiki: Short Description)
Use an open source object detector deep learning model to display and filter objects recognized in an image in a web application

### Code Pattern Description (wiki: Introduction)
The IBM Model Asset eXchange (MAX) has given application developers without data science experience easy access to prebuilt machine learning models. This code pattern shows how to create a simple web application to visualize the text output of a MAX model. The web app uses the Object Detector from MAX and creates a simple web UI that displays bounding boxes around detected objects in an image and lets you filter the objects based on their label and probable accuracy given by the model.

ajbozarth commented:
I just pushed the overview paragraph as well as a handful of other minor changes. I also updated the icons README with the icons' License. I've updated the PR title and description to match these changes and moved this out of WIP.

I hope to merge this by EoD Monday, no later than lunchtime Tuesday, and would appreciate reviews before then. I'll also be submitting the code pattern proposal tonight.

ptitzler commented on README.md:
```diff
@@ -18,7 +46,25 @@ free and open source deep learning models.
+#### Start the Model API
+
+1. [Deploy the Model](#1-deploy-the-model)
+2. [Experimenting with the API (Optional)](#2-experimenting-with-the-api-optional)
```
Consistency: "Experiment with ..."

ptitzler replied:
That's an interesting differentiation but not likely one you'd get any tech writer to buy in. Imagine a translator would have to decide whether something is optional based on that rule - the results could be totally unpredictable. Take a look at some (random) official IBM documentation, such as https://console.bluemix.net/docs/containers/cs_firewall.html#firewall_bx 
Note that all sections start with a verb ending in "ing"). Each step starts with a command (Log in, get, retrieve, ...) and optional steps are preceded by "Optional" (see the first step 7 in above doc link).

ajbozarth replied:
Fair enough, I'll change them

ptitzler commented on README.md:
```diff
@@ -18,7 +46,25 @@ free and open source deep learning models.
+1. [Get a local copy of the repository](#1-get-a-local-copy-of-the-repository)
+2. [Install dependencies](#2-install-dependencies)
+3. [Start the Web App Server](#3-start-the-web-app-server)
+4. [Configuring ports (Optional)](#4-configuring-ports-optional)
```
Consistency: "Configure ..."

ajbozarth replied:
see above

ptitzler commented on README.md:
```diff
@@ -18,7 +46,25 @@ free and open source deep learning models.
+
+#### 1. Deploy the Model
 
 To run the docker image, which automatically starts the model serving API, run:
```
Captialize: "Docker image"

ajbozarth replied:
see above

ptitzler commented on README.md:
```diff
@@ -30,12 +76,73 @@ This will pull a pre-built image from Docker Hub (or use an existing image if al
 If you'd rather build the model locally you can follow the steps in the
 [model README](https://github.com/IBM/MAX-Object-Detector/#steps).
 
+#### 2. Experimenting with the API (Optional)
```
See earlier comment "Experiment ..."

ajbozarth replied:
see above

ptitzler commented on README.md:
```diff
@@ -1,7 +1,35 @@
+The server hosts a client-side web UI and relays API calls to the model from the web UI to a REST
+end point for the model. The Web UI takes in an image and sends it to the model REST endpoint
+via the server and displays the detected objects on the UI. The model's REST endpoint is set up
+using the docker image provided on MAX. The Web UI displays the detected objects in an image
```
Capitalize "Docker ..."

ajbozarth replied:
docker should be lowercase when referring to a "docker image" it's only uppercase when referring to the company or application

ajbozarth replied:
it's consistent across MAX, and Docker docs don't actual use docker mostly because it's all about docker and so the adjective is unnecessary, but I did see it in a few places on the Docker site outside the docs at some point, and IBM Cloud docs seem to follow what I'm doing as well.

ptitzler commented on README.md:
```diff
@@ -18,7 +46,25 @@ free and open source deep learning models.
+0. [Use the embedded web app](#use-the-embedded-web-app)
+1. [Get a local copy of the repository](#1-get-a-local-copy-of-the-repository)
+2. [Install dependencies](#2-install-dependencies)
+3. [Start the Web App Server](#3-start-the-web-app-server)
```
Inconsistent capitalization (Other items in the list are not capitalized).

ptitzler commented on README.md:
```diff
@@ -30,12 +76,73 @@ This will pull a pre-built image from Docker Hub (or use an existing image if al
+
+Use the `model/predict` endpoint to load a test image and get predicted labels for the image from the API.
+The coordinates of the bounding box are returned in the `detection_box` field, and contain the array of normalized
+coordinates (ranging from 0 to 1) in the form `[ymin, xmin, ymax, xmax]`.
```
It would be helpful to add how to translate the normalized coordinates into coordinates that one would use (e.g to draw a box)

ajbozarth replied:
That feature is the point of this code pattern, I don't think explaining the details of how bits of code work in the readme is good documentation for a code pattern.

ptitzler replied:
You are referring to the MAX-Object-Detector repository readme? If you are, then yes, I agree.

ajbozarth replied:
I believe we're on the same page. I have a reference to the model repo's readme for more detailed documentation. As for the having an example for how to utilize the coordinates, that example is this app itself, aka the code.

ptitzler commented on README.md:
```diff
@@ -30,12 +76,73 @@ This will pull a pre-built image from Docker Hub (or use an existing image if al
 
+#### Use the embedded web app
+
 The latest release of the web app is deployed with the model API above and is available at
```
Please add a blurb how the embedded mini-app can be disabled.

ajbozarth replied:
Given this is the web app repo and those instructions are in the model repo I don't see a reason to include those instructions. Why would the user want to turn off the feature they're here to use. And as for the case they plan to deploy the standalone instead, there's no harm or overhead in leaving the web app on vs disabling it.

ptitzler replied:
The reason I brought this up is this sentence in this readme: "The latest release of the web app is deployed with the model API above and is available at http://localhost:5000/app".

ajbozarth replied:
That blurb is specifically to state that the web app is already available from the previous step and they don't have to run the following setup unless they want to run the latest master. So that section is actually the "default instructions" for the code pattern, thus the reason telling how to disable it doesn't make sense

ptitzler commented on README.md:
```diff
@@ -30,12 +76,73 @@ This will pull a pre-built image from Docker Hub (or use an existing image if al
 [`http://localhost:5000/app`](http://localhost:5000/app)
 
-To start the web app using the latest development code follow the steps below:
+To start the web app using Node.js and running the latest code, follow the steps below:
```
In line 163 below (can't tag the line) node.js needs to be capitalized.

ajbozarth replied:
I'll fix that

ajbozarth commented:
@ptitzler I fixed the capitalization issues and responded to you other comments.
